### ISSUE TITLE
Bug - Not validating trailing colon

### ISSUE THREAD

**Issue by pedroivorbgrodrigues** [stylelint/stylelint#2026]
Hello fellow devs,

According to w3c css validator, the snippet below is not valid because of the trailing colon, but the linter gave me no errors. It was a misstype but took me a lot to figure out the css file was the problem not something else. I used the standard configuration on atom linter-stylelint plugin.

```css
.counter: {
  min-width: 100%;
  border: solid 3px red;
}
```

linter-stylelint version: 3.5.0
stylelint: 7.5.0
stylelint-config-standard: ^14.0.0

**Comment by evilebottnawi** [comment]
@pearsonhenri first - it is not valid css (valid `.counter\:`), second - format your issue as template

**Comment by Arcanemagus** [comment]
@evilebottnawi The bug report is that it _isn't_ valid CSS... and Stylelint wasn't reporting that.

**Comment by evilebottnawi** [comment]
@Arcanemagus i don't see that rule should reporting about this. I have said it would be a good idea in the future to add validation rules selectors (allowed characters).

**Comment by evilebottnawi** [comment]
In this case i am propose new rule `selector-no-invalid-characters` (maybe best name) @jeddy3 @davidtheclark that do you think?

**Comment by Arcanemagus** [comment]
Interesting, so it doesn't show as a CssSyntaxError from PostCSS, but it is invalid CSS?

Are you sure this isn't a bug in PostCSS for not throwing a syntax error?

**Comment by evilebottnawi** [comment]
@Arcanemagus good question, maybe you are right and postcss should throw `CssSyntaxError`.
@ai What do you think about it?

**Comment by davidtheclark** [comment]
PostCSS does not validate CSS. It's meant to be flexible because people like their non-standard syntaxes, so it actually tries to accept as much variation as it can without breaking, I think. This colon after the selector is actually in use for non-standard syntax — we've dealt with it before here in stylelint, right? 

@evilebottnawi I don't know how you would plan to implement `selector-no-invalid-characters`. Is it possible?

**Comment by ai** [comment]
Technically, there is a very similar syntax of apply-mixins for it.

So CSS syntax spec allow it. Problems is that current browsers, doesn't **understand** it. But it is a **valid syntax**.

But, sure, we can add some simple rule for it. Like `understandable-selectors`.

**Comment by evilebottnawi** [comment]
@davidtheclark yep, we have allowed characters(tokens) in css specs. I can investigation this.

**Comment by davidtheclark** [comment]
@evilebottnawi The real trouble, though, is that `:` *is* an allowable token. So how do you determine when it's not valid? You'd need to write a selector parser — but we've been pushing for postcss-selector-parser to be *more* flexible, not *less*, in order to allow non-standard syntax.

**Comment by jeddy3** [comment]
In keeping with [`selector-no-empty`](http://stylelint.io/user-guide/rules/selector-no-empty/) (which, similar to this case, dealt with _invalid_ CSS that is parsed by PostCSS), we could create `selector-pseudo-class-no-empty` and `selector-pseudo-element-no-empty` rules?

The former checking for `.selector: {}` and the latter for `.selector:: {}`.

**Comment by evilebottnawi** [comment]
@davidtheclark we can check selector on `isStandardSyntaxSelector` (where process all non standard situation, including non standard syntax and cssnext syntax), when check on invalid characters.

**Comment by evilebottnawi** [comment]
@jeddy3 sound good

**Comment by evilebottnawi** [comment]
@jeddy3 but we can meet later request about invalidate selector `.test? { color: red; }` (or some other characters), i think `no-empty` is not good solution.

**Comment by davidtheclark** [comment]
@jeddy3 @evilebottnawi I'm not sure about this `selector-pseudo-class-no-empty`. It seems like overkill to create a rule to address this. I think the issue isn't really about an empty pseudo-class, either, but a misunderstanding of CSS syntax. What about all the other invalid CSS that can be passed through? 

I'm just not convinced we have the means to enforce "valid" CSS *and* allow non-standard CSS. I think we should close this issue and just acknowledge that.

**Comment by evilebottnawi** [comment]
@jeddy3 Anyone can inadvertently write an incorrect selector, allowing the input of incorrect characters, and it would be more true to create a rule for the validation of this. I think `no-empty` is ambiguous solution. In theory, it is generally incorrect syntax.

**Comment by evilebottnawi** [comment]
@davidtheclark I think it is not necessary to close the discussion, we should decide whether we are going the right way use `no-empty` or not use

**Comment by evilebottnawi** [comment]
@davidtheclark @jeddy3 How about use
> we can check selector on isStandardSyntaxSelector (where process all non standard situation, including non standard syntax and cssnext syntax), when check on invalid characters.

Seem good to me, we exclude all non standard and check on valid characters only selector, i can do experimental PR

**Comment by evilebottnawi** [comment]
Ref: https://drafts.csswg.org/selectors-4/#invalid

**Comment by davidtheclark** [comment]
@evilebottnawi Still not on board. Are you proposing to create a selector parser? I think that's the only way this is going to work.

**Comment by davidtheclark** [comment]
Followup to ^^ — I think creating a selector parser needs to be an external module, and will be a lot of work, and will duplicate work already done.

**Comment by evilebottnawi** [comment]
@davidtheclark not parser, validator (we don't need tokens and ast, i think the regular expression would be enough), about external module i am agree

**Comment by evilebottnawi** [comment]
@davidtheclark in theory, any selector is `ident` ([a-zA-Z0-9-_], the remaining characters require escaping)

**Comment by davidtheclark** [comment]
Pseudo-selectors add lots of possibilities. And anything can be inside a quoted string, right?

**Comment by evilebottnawi** [comment]
Special symbols (example `#` for `id selector`, ':' for pseudo classes and etc) well documented and practically not change, i think `walk` on `selectors` then `switch and case` by `type` and valid ident - it is all rule.
> Pseudo-selectors add lots of possibilities. And anything can be inside a quoted string, right?
```
<pseudo-class-selector> = ':' <ident-token> |
                          ':' <function-token> <any-value> ')'

<pseudo-element-selector> = ':' <pseudo-class-selector>
```
Where `<function-token>` is `<ident-token>(` and `any-value` is any characters.

**Comment by evilebottnawi** [comment]
I do propose to not close the issue. Personally, my opinion as we do linters, is one of the most important rule, the right to a random typo in writing may allow everyone, especially when a lot of code and I would like to see this rule. I also understand that we all have a little time on the new features, but perhaps any living creature shall be taken, and write the rule. The main problem just write a regular expression to validate selector. I just like to put a milestone `future-major` and labels `help wanted`, `rule`. I agree that the validation and testing, should be placed in a **separate repo**, to facilitate support.

**Comment by davidtheclark** [comment]
> my opinion as we do linters, is one of the most important rule, the right to a random typo in writing may allow everyone, especially when a lot of code and I would like to see this rule.

Agreed that it would be nice, but I think SCSS and Less and those tools ruined the opportunity with all their non-standard syntax. Anyway, if you want to try, go for it. But I think you should close this issue and create a new one with a focused purpose.

**Comment by evilebottnawi** [comment]
@davidtheclark let's create new issue with a focused purpose

**Comment by jeddy3** [comment]
Closing this issue.

Sounds like a good candidate for a plugin like [`stylelint-csstree-validator`](https://github.com/csstree/stylelint-validator).
